File: api/routers/groups.py
"""ワークスペースグループの CRUD エンドポイント。

グループは config.json の __global__.groups に保存される。
ワークスペースへの割り当ては workspace エントリの group_id フィールドで管理する。
"""

import logging

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from ..auth import verify_token
from ..common import GLOBAL_CONFIG_KEY, generate_entity_id
from ..config import load_all_config, save_all_config
from ..errors import not_found

logger = logging.getLogger(__name__)

router = APIRouter(dependencies=[Depends(verify_token)])


def _load_groups(all_config: dict) -> list[dict]:
    return list(all_config.get(GLOBAL_CONFIG_KEY, {}).get("groups", []))


def _save_groups(groups: list[dict]) -> None:
    all_config = load_all_config()
    global_section = dict(all_config.get(GLOBAL_CONFIG_KEY, {}))
    global_section["groups"] = groups
    all_config[GLOBAL_CONFIG_KEY] = global_section
    save_all_config(all_config)


class GroupRequest(BaseModel):
    name: str


class GroupOrderRequest(BaseModel):
    order: list[str]


@router.put("/group-order")
def update_group_order(body: GroupOrderRequest):
    all_config = load_all_config()
    groups = _load_groups(all_config)
    order_map = {gid: i for i, gid in enumerate(body.order)}
    sorted_groups = sorted(groups, key=lambda g: order_map.get(g["id"], len(groups)))
    _save_groups(sorted_groups)
    logger.info("group order updated count=%d", len(body.order))
    return [g["id"] for g in sorted_groups]


@router.get("/groups")
def list_groups():
    all_config = load_all_config()
    return _load_groups(all_config)


@router.post("/groups")
def create_group(body: GroupRequest):
    name = body.name.strip()
    if not name:
        from ..errors import bad_request
        raise bad_request("Group name is required")
    group_id = generate_entity_id()
    all_config = load_all_config()
    groups = _load_groups(all_config)
    groups.append({"id": group_id, "name": name})
    _save_groups(groups)
    logger.info("group created id=%s name=%s", group_id, name)
    return {"id": group_id, "name": name}


@router.put("/groups/{group_id}")
def update_group(group_id: str, body: GroupRequest):
    name = body.name.strip()
    if not name:
        from ..errors import bad_request
        raise bad_request("Group name is required")
    all_config = load_all_config()
    groups = _load_groups(all_config)
    for group in groups:
        if group["id"] == group_id:
            group["name"] = name
            _save_groups(groups)
            logger.info("group updated id=%s name=%s", group_id, name)
            return {"id": group_id, "name": name}
    raise not_found(f"Group '{group_id}' not found")


@router.delete("/groups/{group_id}")
def delete_group(group_id: str):
    all_config = load_all_config()
    groups = _load_groups(all_config)
    new_groups = [g for g in groups if g["id"] != group_id]
    if len(new_groups) == len(groups):
        raise not_found(f"Group '{group_id}' not found")
    # グループに所属するワークスペースの group_id をクリアする
    for key, entry in all_config.items():
        if key == GLOBAL_CONFIG_KEY or not isinstance(entry, dict):
            continue
        if entry.get("group_id") == group_id:
            entry["group_id"] = None
    global_section = dict(all_config.get(GLOBAL_CONFIG_KEY, {}))
    global_section["groups"] = new_groups
    all_config[GLOBAL_CONFIG_KEY] = global_section
    save_all_config(all_config)
    logger.info("group deleted id=%s", group_id)
    return {"status": "ok"}

File: server/src/groups.rs
//! ワークスペースグループの CRUD（Python 側 `api/routers/groups.py` の移植）。
//!
//! グループは config.json の `__global__.groups` に保存される。
//! ワークスペースへの割り当ては workspace エントリの group_id フィールドで管理する。

use std::sync::Arc;

use axum::extract::{Path, State};
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Map, Value};

use crate::auth::RequireAuth;
use crate::config::{generate_entity_id, GLOBAL_CONFIG_KEY};
use crate::errors::{bad_request, not_found, ApiError};
use crate::state::AppState;
use crate::util::JsonBody;

fn load_groups(all_config: &Map<String, Value>) -> Vec<Value> {
    all_config
        .get(GLOBAL_CONFIG_KEY)
        .and_then(Value::as_object)
        .and_then(|g| g.get("groups"))
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
}

#[derive(Deserialize)]
pub struct GroupRequest {
    name: String,
}

#[derive(Deserialize)]
pub struct GroupOrderRequest {
    order: Vec<String>,
}

pub async fn list_groups(State(state): State<Arc<AppState>>, _auth: RequireAuth) -> Json<Value> {
    Json(Value::Array(load_groups(&state.config.load_all())))
}

pub async fn create_group(
    State(state): State<Arc<AppState>>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<GroupRequest>,
) -> Result<Json<Value>, ApiError> {
    let name = body.name.trim().to_string();
    if name.is_empty() {
        return Err(bad_request("Group name is required"));
    }
    let group_id = generate_entity_id();
    state.config.with_exclusive(|all_config| {
        let mut groups = load_groups(all_config);
        groups.push(json!({"id": &group_id, "name": &name}));
        crate::config::ConfigStore::merge_global_section(
            all_config,
            "groups",
            Value::Array(groups),
        );
        Ok(())
    })?;
    tracing::info!("group created id={group_id} name={name}");
    Ok(Json(json!({"id": group_id, "name": name})))
}

pub async fn update_group(
    State(state): State<Arc<AppState>>,
    Path(group_id): Path<String>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<GroupRequest>,
) -> Result<Json<Value>, ApiError> {
    let name = body.name.trim().to_string();
    if name.is_empty() {
        return Err(bad_request("Group name is required"));
    }
    state.config.with_exclusive(|all_config| {
        let mut groups = load_groups(all_config);
        let found = groups.iter_mut().any(|group| {
            if group.get("id").and_then(Value::as_str) == Some(group_id.as_str()) {
                group["name"] = Value::String(name.clone());
                true
            } else {
                false
            }
        });
        if !found {
            return Err(not_found(format!("Group '{group_id}' not found")));
        }
        crate::config::ConfigStore::merge_global_section(
            all_config,
            "groups",
            Value::Array(groups),
        );
        Ok(())
    })?;
    tracing::info!("group updated id={group_id} name={name}");
    Ok(Json(json!({"id": group_id, "name": name})))
}

pub async fn delete_group(
    State(state): State<Arc<AppState>>,
    Path(group_id): Path<String>,
    _auth: RequireAuth,
) -> Result<Json<Value>, ApiError> {
    state.config.with_exclusive(|all_config| {
        let groups = load_groups(all_config);
        let new_groups: Vec<Value> = groups
            .iter()
            .filter(|g| g.get("id").and_then(Value::as_str) != Some(group_id.as_str()))
            .cloned()
            .collect();
        if new_groups.len() == groups.len() {
            return Err(not_found(format!("Group '{group_id}' not found")));
        }
        // グループに所属するワークスペースの group_id をクリアする
        for (key, entry) in all_config.iter_mut() {
            if key == GLOBAL_CONFIG_KEY {
                continue;
            }
            let Some(obj) = entry.as_object_mut() else {
                continue;
            };
            if obj.get("group_id").and_then(Value::as_str) == Some(group_id.as_str()) {
                obj.insert("group_id".to_string(), Value::Null);
            }
        }
        crate::config::ConfigStore::merge_global_section(
            all_config,
            "groups",
            Value::Array(new_groups),
        );
        Ok(())
    })?;
    tracing::info!("group deleted id={group_id}");
    Ok(Json(json!({"status": "ok"})))
}

pub async fn update_group_order(
    State(state): State<Arc<AppState>>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<GroupOrderRequest>,
) -> Result<Json<Value>, ApiError> {
    let order_map: std::collections::HashMap<&str, usize> = body
        .order
        .iter()
        .enumerate()
        .map(|(i, gid)| (gid.as_str(), i))
        .collect();
    let ids = state.config.with_exclusive(|all_config| {
        let groups = load_groups(all_config);
        let mut sorted_groups = groups;
        let fallback = sorted_groups.len();
        sorted_groups.sort_by_key(|g| {
            g.get("id")
                .and_then(Value::as_str)
                .and_then(|id| order_map.get(id).copied())
                .unwrap_or(fallback)
        });
        let ids: Vec<Value> = sorted_groups
            .iter()
            .filter_map(|g| g.get("id").cloned())
            .collect();
        crate::config::ConfigStore::merge_global_section(
            all_config,
            "groups",
            Value::Array(sorted_groups),
        );
        Ok(ids)
    })?;
    tracing::info!("group order updated count={}", body.order.len());
    Ok(Json(Value::Array(ids)))
}
